### PR TITLE
Change order of series vs. data check

### LINKED ISSUE #166: Subsampling series slicing is done before asserting series and data equality
In [subsampling.py](https://github.com/alchemistry/alchemlyb/blob/master/src/alchemlyb/preprocessing/subsampling.py#L177) the series is compared to the data **after** already slicing it. This will lead to an error thrown whenever `lower != None`.

### PR DESCRIPTION
**Pull request by schlaicha**
fixes #166

### CODE REVIEW

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thank you for the fix, @schlaicha .

Please also add a test that 
* fails _before_ making your fix
* passes _after_ adding your fix
* 
which will make sure that we do not later introduce a regression. Please ask if you want help — the tests can be a bit confusing at first.

Also add yourself to AUTHORS (start a 2021 section) and add an entry to the 0.6.0 Fixes in CHANGES (you need to pull from your branch to get the updated CHANGES).

**Comment by orbeckst** [comment]
@xiki-tempula could you please review the PR and help if necessary? I did a first look but an in-depth review from you would be good.

**Review by xiki-tempula — CHANGES_REQUESTED** [review]
I would probably put it after the sanity check.

File: src/alchemlyb/preprocessing/subsampling.py [hunk 1]
```diff
@@ -169,13 +169,13 @@ def statistical_inefficiency(df, series=None, lower=None, upper=None, step=None,
             raise KeyError("Statistical inefficiency only works as expected if "
                            "values are sorted by time, increasing.")
 
+    if (len(series) != len(df) or
+        not all(series.reset_index()['time'] == df.reset_index()['time'])):
+        raise ValueError("series and data must be sampled at the same times")
+
     if series is not None:
```
Comment (xiki-tempula) [inline src/alchemlyb/preprocessing/subsampling.py:176]:
```suggestion
    if series is not None:
            if (len(series) != len(df) or
                not all(series.reset_index()['time'] == df.reset_index()['time'])):
                raise ValueError("series and data must be sampled at the same times")
            series = slicing(series, lower=lower, upper=upper, step=step)
               
```

**Comment by schlaicha** [comment]
Thank you for the feedback @orbeckst  and @xiki-tempula ! I will have a look at the tests tomorrow and also why the tests are failing now.

Just for reference, this is why I put my temporary changes into our [flamel PR](https://github.com/alchemistry/flamel/pull/1/files/c7cbfd18cb5e34e9dbab963060b6aadcda7c2b39#r714024154) last year without the intention of ever merging this.

I really appreciate your work on these projects and hope that we can contribute some parts, at least from a user perpective :-)

File: src/alchemlyb/tests/test_preprocessing.py [hunk 2]
```diff
@@ -75,6 +75,16 @@ def test_duplicated_exception(self, data):
...
             self.slicer(data.sort_index(0), lower=200)
 
+    def test_subsample_bounds_and_step(self, gmx_ABFE):
+        """Make sure that slicing the series also works
+        """
+        subsample = statistical_inefficiency(gmx_ABFE,
+                                             gmx_ABFE.sum(axis=1),
+                                             lower=100,
+                                             upper=400,
+                                             step = 2)
```
Comment (orbeckst) [inline src/alchemlyb/tests/test_preprocessing.py:85]: pep8, no spaces
```suggestion
                                             step=2)
```

File: src/alchemlyb/preprocessing/subsampling.py [hunk 3]
```diff
@@ -170,11 +170,12 @@ def statistical_inefficiency(df, series=None, lower=None, upper=None, step=None,
                            "values are sorted by time, increasing.")
 
     if series is not None:
-        series = slicing(series, lower=lower, upper=upper, step=step)
-
+    
         if (len(series) != len(df) or
             not all(series.reset_index()['time'] == df.reset_index()['time'])):
             raise ValueError("series and data must be sampled at the same times")
```
Comment (orbeckst) [inline src/alchemlyb/preprocessing/subsampling.py:176]: Do we test this ValueError anywhere?

Comment (orbeckst) [reply src/alchemlyb/preprocessing/subsampling.py:176]: According to https://app.codecov.io/gh/alchemistry/alchemlyb/compare/167/diff we do, but not sure where.

**Review by orbeckst — APPROVED** [review]
Looks good to me. Thank you.

(I'll fix a minor PEP8 thing via suggestion.)

**Pull request merged**